### PR TITLE
LocalCallExecutor and RemoteCallRequest generic over Hasher and NodeCodec

### PR DESCRIPTION
**Pull request by dvdplm**

### CODE REVIEW

**Comment by gavofyork** [comment]
doesn't build

File: substrate/runtime-io/with_std.rs [hunk 1]
```diff
@@ -102,7 +102,7 @@ pub fn storage_root() -> H256 {
 
 /// A trie root formed from the enumerated items.
 pub fn enumerated_trie_root(serialised_values: &[&[u8]]) -> [u8; 32] {
```
Comment (gavofyork) [inline substrate/runtime-io/with_std.rs]: this should be prefixed with `keccak_` to make its function specific. (i.e. renamed to `keccak_enumerated_trie_root`).

**Comment by gavofyork** [comment]
one thing that might be worth a try: just compiling blake2, twox and the trie functions into wasm and foregoing the extern stubs for them. it'll result in slower and bigger wasm for sure, but since we generally execute with the inbuilt native anyway that's probably not so important. it should make the code quite a lot cleaner.

we should also dispense with using RLP as the data serialisation format for the trienodes - it's overkill, especially if we compile into wasm.

File: substrate/runtime-io/with_std.rs (hunk 1 above)
Comment (dvdplm) [reply substrate/runtime-io/with_std.rs]: I thought the plan was to switch to using Blake2b soon – next 30 - 60 days? – and unless we have multiple crypto/format combos, is it really useful to have them explicitly named?

**Comment by rphmeier** [comment]
> we should also dispense with using RLP as the data serialisation format for the trienodes - it's overkill, especially if we compile into wasm.

The `NodeCodec` trait is meant to do that, but I think we would need to write one for our substrate-codec.

**Comment by gavofyork** [comment]
`LocalCallExecutor` shouldn't be generic over hash or node codec. The only parts of the code that are generic over trie type is within the runtime.

File: demo/executor/src/lib.rs [hunk 2]
```diff
@@ -239,7 +239,7 @@ mod tests {
 			UncheckedExtrinsic::new(extrinsic, signature)
 		}).collect::<Vec<_>>();
 
-		let extrinsics_root = ordered_trie_root(extrinsics.iter().map(Encode::encode)).0.into();
+		let extrinsics_root = ordered_trie_root::<KeccakHasher, _, _>(extrinsics.iter().map(Encode::encode)).0.into();
```
Comment (gavofyork) [inline demo/executor/src/lib.rs]: Shouldn't this also define the `NodeCodec`?

File: substrate/client/db/src/lib.rs [hunk 3]
```diff
@@ -87,10 +87,10 @@ pub fn new_client<E, S, Block>(
 	executor: E,
 	genesis_storage: S,
 	execution_strategy: ExecutionStrategy,
-) -> Result<client::Client<Backend<Block>, client::LocalCallExecutor<Backend<Block>, E>, Block>, client::error::Error>
+) -> Result<client::Client<Backend<Block>, client::LocalCallExecutor<Backend<Block>, E, KeccakHasher, RlpCodec>, Block>, client::error::Error>
```
Comment (gavofyork) [inline substrate/client/db/src/lib.rs]: LocalCallExecutor shouldn't be generic over these args.

File: demo/executor/src/lib.rs (hunk 2 above)
Comment (dvdplm) [reply demo/executor/src/lib.rs]: Ideally yes, but `triehash` is intrinsically tied to RLP and untying that amounts to, I believe, a rewrite. [Issue](https://github.com/paritytech/parity-common/issues/50).

File: substrate/executor/src/lib.rs [hunk 4]
```diff
@@ -76,15 +76,15 @@ pub use native_executor::{with_native_environment, NativeExecutor, NativeExecuti
...
 
 /// Provides runtime information.
-pub trait RuntimeInfo {
+pub trait RuntimeInfo<H: Hasher> {
 	/// Native runtime information if any.
 	const NATIVE_VERSION: Option<RuntimeVersion>;
 
 	/// Extract RuntimeVersion of given :code block
-	fn runtime_version<E: Externalities<KeccakHasher>> (
+	fn runtime_version<E: Externalities<H>> (
```
Comment (gavofyork) [inline substrate/executor/src/lib.rs]: why is this generic?

**Comment by gavofyork** [comment]
`Externalities` appears to be generic over hasher - why?

File: substrate/test-client/src/lib.rs [hunk 5]
```diff
@@ -53,7 +53,10 @@ pub use local_executor::LocalExecutor;
 pub type Backend = client::in_mem::Backend<runtime::Block, KeccakHasher, RlpCodec>;
 
 /// Test client executor.
-pub type Executor = client::LocalCallExecutor<Backend, executor::NativeExecutor<LocalExecutor>>;
+pub type Executor = client::LocalCallExecutor<
+		Backend,
+		executor::NativeExecutor<LocalExecutor>,
+	>;
```
Comment (gavofyork) [inline substrate/test-client/src/lib.rs]: superfluous indent

**Comment by gavofyork** [comment]
Next step is to expunge Keccak and RLP from the codebase!

**Comment by gavofyork** [comment]
lock and binaries again ;-)

**Pull request merged**